### PR TITLE
NIFI-12053 - Update Github CI Builds for 2.x line to Java 21 images Update NiFi Docs/Readme to reference Java 21 as basis

### PR DESCRIPTION
**Pull request by joewitt**
# Summary

[NIFI-00000](https://issues.apache.org/jira/browse/NIFI-00000)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by joewitt** [comment]
The only commit necessary here is the NIFI-12053 commit. The other two are just place holders as those commits are on other PRs waiting to be merged.  But with those and the commit for NIFI-12053 we are fully moved to Java 21 for the entire build process in Github and the full build with contrib check and all profiles is able to work.  NiFi also runs in Java 21.

**Comment by joewitt** [comment]
This change encapsulates any needed --add-opens command line values in the maven build itself or in the nifi bootstrap conf as needed so users dont need to add anything.  It also only applies this during the build for surefire and only in the offending pom(s) which as of now is Hive3 processors.

File: .github/workflows/ci-workflow.yml [hunk 1]
```diff
@@ -97,11 +96,11 @@ jobs:
             ~/.m2/repository
           # Cache Maven modules using a cache key different from setup-java steps
           key: ${{ runner.os }}-maven-static-analysis-${{ hashFiles('**/pom.xml') }}
-      - name: Set up Java 17
+      - name: Set up Java 21
         uses: actions/setup-java@v3
         with:
           distribution: 'zulu'
-          java-version: '17'
+          java-version: '21-ea'
```
Comment (exceptionfactory) [inline .github/workflows/ci-workflow.yml:103]: We should avoid switching an Early Access build version and wait until the official release version is ready.

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/pom.xml [hunk 2]
```diff
@@ -412,6 +412,15 @@
     </dependencies>
     <build>
         <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <environmentVariables>
+                        <JDK_JAVA_OPTIONS>--add-opens java.base/java.net=ALL-UNNAMED</JDK_JAVA_OPTIONS>
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/pom.xml:420]: A comment on the reason for this inclusion avoid be helpful, mentioning the offending Hive class.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/bootstrap.conf [hunk 3]
```diff
@@ -90,6 +85,13 @@ java.arg.17=-Dzookeeper.admin.enableServer=false
 #            org.apache.jasper.servlet.JasperLoader,org.jvnet.hk2.internal.DelegatingClassLoader,org.apache.nifi.nar.NarClassLoader
 # End of Java Agent config for native library loading.
 
+# The following entry is needed in Java 21 because some libraries invoke
+# reflective calls that Java no longer considers allowed by default.
+# https://docs.oracle.com/en/java/javase/16/migrate/migrating-jdk-8-later-jdk-releases.html#GUID-7BB28E4D-99B3-4078-BDC4-FC24180CE82B
+# This may need to be modified if additional reflective access is needed by certain libraries
+# This is only known to be needed for the Hive3 processors as of now.
+java.arg.20=--add-opens=java.base/java.net=ALL-UNNAMED
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/bootstrap.conf:93]: Although this is an unfortunate requirement for Hive 3, the comment is helpful. Hopefully we can remove this in the future with newer Hive versions.

**Review by exceptionfactory — COMMENTED** [review]
Great work pulling together the final changes for Java 21 @joewitt! The majority of the changes look good.

One important detail on timing, I think we should wait until the official release of Java 21 and not require building on an Early Access version. With the release date set for September 19, that is only a couple days away. Once that is available, updating the Java version references for the automated builds should make this ready.

**Comment by ChrisSamo632** [comment]
Docker images will need updating to build & run with Java 21 (once the base images are available) - is this being covered under a separate future ticket?

**Comment by joewitt** [comment]
@ChrisSamo632 Correct that will be covered in another JIRA which is filed https://issues.apache.org/jira/browse/NIFI-12074

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/pom.xml (hunk 2 above)
Comment (joewitt) [reply nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/pom.xml:420]: added comment

**Review by exceptionfactory — APPROVED** [review]
Thanks for making adjustments @joewitt, the latest version looks good with Zulu and Corretto releases of Java 21! +1 merging

**Pull request closed without merging**